src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -1144,10 +1144,24 @@ private void VisitReceiverBeforeCall(BoundExpression receiverOpt, MethodSymbol m
 
         private void VisitReceiverAfterCall(BoundExpression receiverOpt, MethodSymbol method)
         {
-            NamedTypeSymbol containingType;
-            if (receiverOpt != null && ((object)method == null || method.MethodKind == MethodKind.Constructor || (object)(containingType = method.ContainingType) != null && method.RequiresInstanceReceiver && !containingType.IsReferenceType && !TypeIsImmutable(containingType)))
+            if (receiverOpt == null)
             {
-                WriteArgument(receiverOpt, method?.MethodKind == MethodKind.Constructor ? RefKind.Out : RefKind.Ref, method);
+                return;
+            }
+
+            if (method == null)
+            {
+                WriteArgument(receiverOpt, RefKind.Ref, method: null);
```

Comment:
I'm not sure what it means for `method` to be null but adding a `Debug.Assert(method != null)` made a lot of tests fail. This refactor is meant to just preserve the existing behavior as much as possible while deferring to the `ThisParameterSymbol.RefKind` to actually figure out whether to note a receiver write.